### PR TITLE
aerospace: init at 0.14.2-Beta

### PR DESCRIPTION
**Pull request by alexandru0-dev**
## Description of changes
Introduces [aerospace](https://github.com/nikitabobko/AeroSpace) - an i3-like tiling window manager for macOS.
As of now uses binaries as are already signed and building would result in losing accessibility permissions between updates. (see https://github.com/NixOS/nixpkgs/pull/299099#issuecomment-2029566383)

I would like some feedback about man pages and shell-completion if possible.

Closes #299099.
Reason: There wasn't really a package made and unmaintained.
## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ae/aerospace/package.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+in
+stdenv.mkDerivation {
+  inherit pname;
+
+  version = lib.strings.toLower ghReleaseVersion;
+
+  src = fetchzip {
+    url = "${ghUrl}/releases/download/v${ghReleaseVersion}/AeroSpace-v${ghReleaseVersion}.zip";
+    hash = "sha256-v2D/IV9Va0zbGHEwSGt6jvDqQYqha290Lm6u+nZTS3A=";
+  };
```
Comment (SigmaSquadron) [inline pkgs/by-name/ae/aerospace/package.nix:26]: Use `fetchFromGitHub` instead, and drop the `ghUrl` stuff; `passthru.gitUpdater` will work fine with `fetchFromGitHub`.

File: pkgs/by-name/ae/aerospace/package.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  installPhase = ''
+    mkdir -p $out/Applications
+    mv ${appName} $out/Applications
+    cp -R bin $out
+    mkdir -p $out/share
+    cp -R shell-completion $out/share
+    cp -R manpage $out/share/man
+  '';
```
Comment (SigmaSquadron) [inline pkgs/by-name/ae/aerospace/package.nix:35]: I'm not very familiar with Darwin stuff, but are you sure none of these steps are done automatically by `stdenv`?

File: pkgs/by-name/ae/aerospace/package.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  passthru.updateScript = gitUpdater {
+    url = "${ghUrl}.git";
+    rev-prefix = "v";
+  };
+
+  meta = {
+    license = lib.licenses.mit;
+    mainProgram = "aerospace";
+    homepage = "https://github.com/nikitabobko/AeroSpace";
+    description = "An i3-like tiling window manager for macOS";
```
Comment (SigmaSquadron) [inline pkgs/by-name/ae/aerospace/package.nix:48]:
```suggestion
    description = "i3-like tiling window manager for macOS";
```

File: pkgs/by-name/ae/aerospace/package.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    platforms = lib.platforms.darwin;
+    maintainers = with lib.maintainers; [ alexandru0-dev ];
+    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
```
Comment (SigmaSquadron) [inline pkgs/by-name/ae/aerospace/package.nix:51]:
```suggestion
    sourceProvenance = [ lib.sourceTypes.binaryNativeCode ];
```

**Review by SigmaSquadron — COMMENTED** [review]
Overall this looks pretty good! Nice work!

File: pkgs/by-name/ae/aerospace/package.nix (hunk 2 above)
Comment (alexandru0-dev) [reply pkgs/by-name/ae/aerospace/package.nix:35]: Some of them like bin yes, tho man pages and shell autocomplete no bc their name is not the one expected

I could modify the src so they are automatically detected

File: pkgs/by-name/ae/aerospace/package.nix (hunk 1 above)
Comment (alexandru0-dev) [reply pkgs/by-name/ae/aerospace/package.nix:26]: Ohh thx, I didn't know (first time using gitUpdater)

File: pkgs/by-name/ae/aerospace/package.nix (hunk 2 above)
Comment (SigmaSquadron) [reply pkgs/by-name/ae/aerospace/package.nix:35]: Nah, it's alright like this.

**Comment by alexandru0-dev** [comment]
@SigmaSquadron I got gitUpdater working but fails to patch automatically (as it's using the `toLower` function) and making the version static with lowercase makes the patch work but revs are case-sensitive.

Is it ok to break convention here or how should I proceed?

**Comment by SigmaSquadron** [comment]
You haven't pushed your changes so I'm not exactly sure what you've done. Regardless, `r-ryantm` should automatically fetch new versions if it detects the GitHub repo, so you might even be able to omit the `updateScript`.

**Comment by alexandru0-dev** [comment]
@SigmaSquadron done, my only concern would be the capitalization of "Beta" as with the gitUpdater the new proposed version wasn't all lowercase

File: pkgs/by-name/ae/aerospace/package.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    rev = "v${version}";
+    hash = "sha256-fMsw4Oos7ddcvr/agAN3KZMJ8o8N9UTIqipR+qF8jAk=";
+  };
+
+  installPhase = ''
+    mkdir -p $out/Applications
+    mv ${appName} $out/Applications
+    cp -R bin $out
+    mkdir -p $out/share
+    cp -R shell-completion $out/share
```
Comment (wattmto) [inline pkgs/by-name/ae/aerospace/package.nix:34]:
```suggestion
    installShellCompletion --bash shell-completion/bash/aerospace
    installShellCompletion --fish shell-completion/fish/aerospace.fish
    installShellCompletion --zsh shell-completion/zsh/_aerospace
```

File: pkgs/by-name/ae/aerospace/package.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    cp -R manpage $out/share/man
```
Comment (wattmto) [inline pkgs/by-name/ae/aerospace/package.nix:35]:
```suggestion
    installManPage manpage/*
```

File: pkgs/by-name/ae/aerospace/package.nix [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  src = fetchFromGitHub {
+    owner = "nikitabobko";
+    repo = "aerospace";
+    rev = "v${version}";
+    hash = "sha256-fMsw4Oos7ddcvr/agAN3KZMJ8o8N9UTIqipR+qF8jAk=";
+  };
+
+  installPhase = ''
+    mkdir -p $out/Applications
```
Comment (wattmto) [inline pkgs/by-name/ae/aerospace/package.nix:30]:
```suggestion
    runHook preInstall
    mkdir -p $out/Applications
```

File: pkgs/by-name/ae/aerospace/package.nix [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    mv ${appName} $out/Applications
+    cp -R bin $out
+    mkdir -p $out/share
+    cp -R shell-completion $out/share
+    cp -R manpage $out/share/man
+  '';
```
Comment (wattmto) [inline pkgs/by-name/ae/aerospace/package.nix:36]:
```suggestion
  runHook postInstall
  '';
```

**Review by wattmto — CHANGES_REQUESTED** [review]
Thank you for your contribution!

**Comment by alexandru0-dev** [comment]
@wattmto thxx, just rebased
tho I need to "fix" the updateScript bc as of now `gitUpdater` is incapable of recognizing to update the `ghVersion` var and fails
this can be solved using `Beta` (allowing uppercase in the version) instead of trying to force it lowercase
any guidance?

File: pkgs/by-name/ae/aerospace/package.nix (hunk 1 above)
Comment (alexandru0-dev) [reply pkgs/by-name/ae/aerospace/package.nix]: fetchFromGithub breaks the install phase as i'm using the binaries

Reverting to old behaviour

**Comment by SigmaSquadron** [comment]
On 2024-09-26 @ 08:02 PM +10:00 UTC, Alexandru Nechita ***@***.***> wrote:
> @wattmto <https://github.com/wattmto> thxx, just rebased
> tho I need to "fix" the updateScript bc as of now `gitUpdater` is 
> incapable of recognizing to update the `ghVersion` var and fails
> this can be solved using `Beta` (allowing uppercase in the version) 
> instead of trying to force it lowercase
> any guidance?

Just use a capital in the version name. There are a few other packages that do this.

Also, if the zip file being fetched isn't source code to be built, but a pre-compiled package, the sourceProvenance should be adjusted accordingly: https://nixos.org/manual/nixpkgs/stable/#sec-meta-sourceProvenance

**Comment by alexandru0-dev** [comment]
@SigmaSquadron the sourceProvenance was already set to `binaryNativeCode`

just fixed the last things and it's ready for review if you want @wattmto @SigmaSquadron

**Comment by SigmaSquadron** [comment]
On 2024-09-26 @ 08:55 PM +10:00 UTC, Alexandru Nechita ***@***.***> wrote:
> @SigmaSquadron <https://github.com/SigmaSquadron> the sourceProvenance 
> was already set to `binaryNativeCode`

My bad! I didn't notice.

**Comment by alexandru0-dev** [comment]
no worries
just need to remove a line for `nixfmt-check`

**Review by wattmto — APPROVED** [review]
LGTM

**Review by SigmaSquadron — APPROVED** [review]

**Comment by NickHu** [comment]
Thanks for your contribution!

**Pull request merged**